File: backend/utils/emotion_mapping.py
"""
EmotionMapping - 感情マッピングユーティリティ

VRMキャラクターの表情・アニメーション制御のための感情マッピングを提供します。
Mastra版の`emotion-mapping.ts`を参考にしたPython実装。

参考:
- frontend/src/lib/emotion-mapping.ts
"""

from typing import Dict, Literal, Any, List, Optional

# サポートされているVRM表情タイプ (VRM1.0準拠)
SupportedExpression = Literal["neutral", "happy", "sad", "angry", "surprised", "relaxed"]


class EmotionMapping:
    """
    感情マッピングクラス

    様々な感情エイリアスを標準VRM表情にマッピングし、
    アニメーション情報を提供します。表情の強度はマップでは持たず、常に 1.0 を用います。
    """

    # 感情タグ → VRM表情（SupportedExpression）
    EXPRESSION_MAP: Dict[str, SupportedExpression] = {
        "happy": "happy",
        "excited": "happy",
        "sad": "sad",
        "angry": "angry",
        "relaxed": "relaxed",
        "surprised": "surprised",
        "apologetic": "sad",
        "informative": "relaxed",
        "guiding": "relaxed",
        "helpful": "happy",
        "neutral": "neutral",
        "serious": "sad",
        "confident": "relaxed",
        "grateful": "happy",
        "confused": "sad",
    }

    DEFAULT_EXPRESSION: SupportedExpression = "neutral"
    DEFAULT_INTENSITY: float = 1.0
    DEFAULT_EXPRESSION_DURATION: float = 2.0

    # VRM表情→アニメーションマッピングテーブル
    ANIMATION_MAP: Dict[SupportedExpression, str] = {
        "neutral": "idle",
        "happy": "greeting",
        "sad": "thinking",
        "angry": "explaining",
        "surprised": "greeting",
        "relaxed": "thinking",
    }

    DEFAULT_ANIMATION: str = "idle"

    # Visemeマッピング（リップシンク用の口形状）
    # VRMのBlendShape名に対応: A→aa, I→ih, U→ou, E→ee, O→oh, Closed→neutral
    VISEME_MAPPING: Dict[str, str] = {
        "A": "aa",
        "I": "ih",
        "U": "ou",
        "E": "ee",
        "O": "oh",
        "Closed": "neutral",
    }

    # Visemeのtransition時間（口の動きは素早く）
    VISEME_TRANSITION_DURATION: float = 0.1

    @classmethod
    def map_to_expression(cls, emotion: str) -> SupportedExpression:
        """
        任意の感情文字列をサポートされているVRM表情にマッピング

        Args:
            emotion: 感情タグ（"happy", "sad", "neutral", "helpful" 等）

        Returns:
            サポートされているVRM表情 (SupportedExpression)
            サポート外の場合は "neutral" を返す

        Examples:
            >>> EmotionMapping.map_to_expression("happy")
            'happy'
            >>> EmotionMapping.map_to_expression("excited")
            'happy'
            >>> EmotionMapping.map_to_expression("unknown")
            'neutral'
        """
        if not emotion or not isinstance(emotion, str):
            return "neutral"

        normalized_emotion = emotion.lower().strip()
        return cls.EXPRESSION_MAP.get(normalized_emotion, "neutral")

    @classmethod
    def get_animation_for_emotion(cls, emotion: str) -> str:
        """
        感情に対応するアニメーション名を取得

        Args:
            emotion: 感情タグ

        Returns:
            アニメーション名

        Examples:
            >>> EmotionMapping.get_animation_for_emotion("happy")
            'greeting'
            >>> EmotionMapping.get_animation_for_emotion("sad")
            'thinking'
        """
        expression = cls.map_to_expression(emotion)
        return cls.ANIMATION_MAP.get(expression, "idle")

    @classmethod
    def get_intensity_for_emotion(cls, emotion: str) -> float:
        """
        表情強度を返す（EXPRESSION_MAP とは独立し、常に固定値）。

        Args:
            emotion: 感情タグ（互換のため受け取るが強度には影響しない）

        Returns:
            常に ``DEFAULT_INTENSITY``（1.0）

        Examples:
            >>> EmotionMapping.get_intensity_for_emotion("happy")
            1.0
            >>> EmotionMapping.get_intensity_for_emotion("neutral")
            1.0
        """
        return cls.DEFAULT_INTENSITY

    @classmethod
    def get_expression_with_intensity(cls, emotion: str) -> Dict[str, Any]:
        """
        感情タグからVRM表情と強度を一度に取得

        表情は ``EXPRESSION_MAP``／``map_to_expression`` に従い、強度は常に ``DEFAULT_INTENSITY``。

        Args:
            emotion: 感情タグ

        Returns:
            VRM表情と強度を含む辞書
            {
                "expression": str,  # VRM表情（"neutral", "happy", "sad"等）
                "intensity": float   # 常に 1.0（DEFAULT_INTENSITY）
            }

        Examples:
            >>> result = EmotionMapping.get_expression_with_intensity("helpful")
            >>> result["expression"]
            'happy'
            >>> result["intensity"]
            1.0
            >>> result = EmotionMapping.get_expression_with_intensity("happy")
            >>> result["expression"]
            'happy'
            >>> result["intensity"]
            1.0
        """
        return {
            "expression": cls.map_to_expression(emotion),
            "intensity": cls.DEFAULT_INTENSITY,
        }

    @classmethod
    def is_supported_emotion(cls, emotion: str) -> bool:
        """
        感情がサポートされているかチェック

        Args:
            emotion: 感情タグ

        Returns:
            サポートされている場合True

        Examples:
            >>> EmotionMapping.is_supported_emotion("happy")
            True
            >>> EmotionMapping.is_supported_emotion("unknown")
            False
        """
        if not emotion or not isinstance(emotion, str):
            return False
        return emotion.lower() in cls.EXPRESSION_MAP

    @classmethod
    def get_supported_expressions(cls) -> list[SupportedExpression]:
        """
        サポートされているすべてのVRM表情を取得

        Returns:
            サポートされているVRM表情のリスト

        Examples:
            >>> expressions = EmotionMapping.get_supported_expressions()
            >>> "neutral" in expressions
            True
            >>> "happy" in expressions
            True
        """
        return ["neutral", "happy", "sad", "angry", "surprised", "relaxed"]

    @classmethod
    def get_all_emotion_aliases(cls) -> list[str]:
        """
        すべての感情エイリアスを取得

        Returns:
            感情エイリアスのリスト

        Examples:
            >>> aliases = EmotionMapping.get_all_emotion_aliases()
            >>> "happy" in aliases
            True
            >>> "excited" in aliases
            True
        """
        return list(cls.EXPRESSION_MAP.keys())

    @classmethod
    def normalize_emotion(
        cls, emotion: str, intensity: float | None = None
    ) -> dict[str, SupportedExpression | float]:
        """
        感情を正規化（VRM表情と強度を返す）

        Args:
            emotion: 感情タグ
            intensity: 互換のため残すが無視される（強度は常に ``DEFAULT_INTENSITY``）

        Returns:
            正規化されたVRM表情と強度の辞書
            {
                "expression": SupportedExpression,
                "intensity": float  # 常に DEFAULT_INTENSITY
            }

        Examples:
            >>> result = EmotionMapping.normalize_emotion("happy")
            >>> result["expression"]
            'happy'
            >>> result["intensity"]
            1.0
            >>> result = EmotionMapping.normalize_emotion("happy", 0.5)
            >>> result["intensity"]
            1.0
        """
        return {
            "expression": cls.map_to_expression(emotion),
            "intensity": cls.DEFAULT_INTENSITY,
        }

    @classmethod
    def extract_viseme_from_lipsync_data(
        cls, lipsync_data: List[Dict[str, Any]]
    ) -> Optional[Dict[str, Any]]:
        """
        リップシンクデータからViseme式を抽出

        リップシンクデータの最初のフレームから口形状情報を取得し、
        VRM制御用のViseme式オブジェクトを生成します。

        Args:
            lipsync_data: リップシンクデータのリスト
                [
                    {
                        "time": float,
                        "volume": float,
                        "mouthOpen": float,
                        "mouthShape": str  # "A", "I", "U", "E", "O", "Closed"
                    },
                    ...
                ]

        Returns:
            Viseme式オブジェクト（リップシンクデータが無効な場合はNone）
            {
                "name": str,  # Viseme名（"aa", "ih", "ou", "ee", "oh", "neutral"）
                "value": float,  # 口の開き具合（0.0-1.0）
                "transition": float  # 遷移時間（秒）
            }

        Examples:
            >>> lipsync_data = [
            ...     {"time": 0.0, "volume": 0.5, "mouthOpen": 0.3, "mouthShape": "A"}
            ... ]
            >>> viseme = EmotionMapping.extract_viseme_from_lipsync_data(lipsync_data)
            >>> viseme["name"]
            'aa'
            >>> viseme["value"]
            0.3
        """
        if not lipsync_data or len(lipsync_data) == 0:
            return None

        try:
            # 最初のフレームからmouthShapeとmouthOpenを取得
            first_frame = lipsync_data[0]
            mouth_shape = first_frame.get("mouthShape", "Closed")
            mouth_open = first_frame.get("mouthOpen", 0.0)

            # mouth_openの値を検証（0.0-1.0の範囲に制限）
            if not isinstance(mouth_open, (int, float)):
                mouth_open = 0.0
            else:
                mouth_open = max(0.0, min(1.0, float(mouth_open)))

            # Visemeマッピングを適用
            viseme_name = cls.VISEME_MAPPING.get(mouth_shape, "neutral")

            # Viseme式オブジェクトを生成
            return {
                "name": viseme_name,
                "value": mouth_open,
                "transition": cls.VISEME_TRANSITION_DURATION,
            }

        except Exception:
            # エラー時はNoneを返す（呼び出し側でエラーハンドリング）
            return None

File: frontend/src/lib/emotion-mapping.ts
/**
 * Centralized emotion mapping for VRM character expressions and animations.
 * Kept in sync with backend/utils/emotion_mapping.py (EXPRESSION_MAP, ANIMATION_MAP).
 *
 * @see backend/utils/emotion_mapping.py
 */

/** VRM 表情（Python SupportedExpression と同一） */
export type SupportedExpression =
  | 'neutral'
  | 'happy'
  | 'sad'
  | 'angry'
  | 'surprised'
  | 'relaxed';

/** public/animations/manifest.json の .vrma 一覧（Python SupportedAnimation と同一） */
export type SupportedAnimation =
  | 'idle'
  | 'bowing'
  | 'greeting'
  | 'looking'
  | 'surprised'
  | 'talking'
  | 'thinking';

/** VRM 表情 6 種（バックエンド get_supported_expressions と同順） */
export const CANONICAL_EXPRESSIONS: readonly SupportedExpression[] = [
  'neutral',
  'happy',
  'sad',
  'angry',
  'surprised',
  'relaxed',
] as const;

export interface EmotionMappingConfig {
  /** 感情タグ → VRM 表情 */
  expressionMap: Record<string, SupportedExpression>;
  /** 感情タグ → アニメーション（感情キーで参照、表情経由ではない） */
  animationMap: Record<string, SupportedAnimation>;
  /** VRM 表情ごとの強度（バックエンドと同様 1.0 固定） */
  intensities: Record<SupportedExpression, number>;
}

export class EmotionMapping {
  /** emotion → VRM 表情（backend EXPRESSION_MAP） */
  static readonly EXPRESSION_MAP: Record<string, SupportedExpression> = {
    happy: 'happy',
    excited: 'happy',
    sad: 'sad',
    angry: 'angry',
    relaxed: 'relaxed',
    surprised: 'surprised',
    apologetic: 'sad',
    informative: 'relaxed',
    guiding: 'relaxed',
    helpful: 'happy',
    neutral: 'neutral',
    serious: 'sad',
    confident: 'relaxed',
    grateful: 'happy',
    confused: 'sad',
  };

  /** emotion → SupportedAnimation（backend ANIMATION_MAP） */
  static readonly ANIMATION_MAP: Record<string, SupportedAnimation> = {
    neutral: 'idle',
    happy: 'greeting',
    excited: 'greeting',
    sad: 'thinking',
    angry: 'talking',
    relaxed: 'thinking',
    surprised: 'surprised',
    apologetic: 'thinking',
    informative: 'talking',
    guiding: 'talking',
    helpful: 'greeting',
    serious: 'thinking',
    confident: 'talking',
    grateful: 'greeting',
    confused: 'thinking',
  };

  static readonly DEFAULT_EXPRESSION: SupportedExpression = 'neutral';
  static readonly DEFAULT_ANIMATION: SupportedAnimation = 'idle';
  static readonly DEFAULT_INTENSITY = 1.0;

  /**
   * 任意の感情文字列をサポート VRM 表情へ（backend map_to_expression）
   */
  static mapToExpression(emotion: string | unknown): SupportedExpression {
    if (!emotion || typeof emotion !== 'string') {
      console.warn('[EmotionMapping] Invalid emotion:', emotion, 'type:', typeof emotion);
      return this.DEFAULT_EXPRESSION;
    }
    const normalized = emotion.toLowerCase().trim();
    return this.EXPRESSION_MAP[normalized] ?? this.DEFAULT_EXPRESSION;
  }

  /**
   * 感情タグからアニメーション名（backend get_animation_for_emotion）
   * ANIMATION_MAP を expression キーで直接参照する。
   */
  static mapToAnimation(emotion: string): SupportedAnimation {
    if (!emotion || typeof emotion !== 'string') {
      return this.DEFAULT_ANIMATION;
    }
    const normalized = emotion.toLowerCase().trim();
    return this.ANIMATION_MAP[normalized] ?? this.DEFAULT_ANIMATION;
  }

  /**
   * 表情強度（backend get_intensity_for_expression — 常に固定値）
   */
  static getIntensityForExpression(_expression: string): number {
    return this.DEFAULT_INTENSITY;
  }
  
  static getSupportedExpressions(): SupportedExpression[] {
    return [...CANONICAL_EXPRESSIONS];
  }

  static getConfig(): EmotionMappingConfig {
    const intensities = {} as Record<SupportedExpression, number>;
    for (const e of CANONICAL_EXPRESSIONS) {
      intensities[e] = this.DEFAULT_INTENSITY;
    }
    return {
      expressionMap: { ...this.EXPRESSION_MAP },
      animationMap: { ...this.ANIMATION_MAP },
      intensities,
    };
  }

  /**
   * 正規化（backend normalize_emotion — intensity 引数は互換のため無視し常に 1.0）
   */
  static normalizeExpression(emotion: string, _intensity?: number): {
    expression: SupportedExpression;
    intensity: number;
  } {
    return {
      expression: this.mapToExpression(emotion),
      intensity: this.DEFAULT_INTENSITY,
    };
  }
}
